timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+            cur_frames = all_frames[lamb_idx]
+
+            prev_boxes = all_boxes[lamb_idx - 1]
+            cur_boxes = all_boxes[lamb_idx]
+
+            prev_U_fns = all_batch_U_fns[lamb_idx - 1]
+            cur_U_fns = all_batch_U_fns[lamb_idx]
+
+            # loop over bond, angle, torsion, nonbonded terms etc.
+            all_fwd_delta_us = []
+            all_rev_delta_us = []
+
+            for u_idx, (prev_U_fn, cur_U_fn) in enumerate(zip(prev_U_fns, cur_U_fns)):
+                fwd_delta_u = beta * (cur_U_fn(prev_frames, prev_boxes) - prev_U_fn(prev_frames, prev_boxes))
+                rev_delta_u = beta * (prev_U_fn(cur_frames, cur_boxes) - cur_U_fn(cur_frames, cur_boxes))
+
+                dG, dG_err = pymbar.BAR(fwd_delta_u, rev_delta_u)
+                dG /= beta
+                dG_err /= beta
+
+                plot_BAR(dG, dG_err, fwd_delta_u, rev_delta_u, U_names[u_idx], get_plot_idx())
+
+                all_fwd_delta_us.append(fwd_delta_u)
+                all_rev_delta_us.append(rev_delta_u)
+
+            # sanity check - I don't think the dG calculation commutes with its components, so we have to re-estimate
+            # the dG from the sum of the delta_us as opposed to simply summing the component dGs
+            total_fwd_delta_us = np.sum(all_fwd_delta_us, axis=0)
+            total_rev_delta_us = np.sum(all_rev_delta_us, axis=0)
+            total_dG, total_dG_err = pymbar.BAR(total_fwd_delta_us, total_rev_delta_us)
+            total_dG /= beta
+            total_dG_err /= beta
+
+            plot_BAR(
+                total_dG,
+                total_dG_err,
+                total_fwd_delta_us,
+                total_rev_delta_us,
+                f"{prefix}_{lamb_idx-1}_to_{lamb_idx}",
+                get_plot_idx(),
+            )
+
+            all_dGs.append(total_dG)
+            all_errs.append(total_dG_err)
+
+            print(
+                f"{prefix} BAR: lambda {lambda_schedule[lamb_idx-1]:.3f} -> {lambda_schedule[lamb_idx]:.3f} dG: {total_dG:.3f} +- {total_dG_err:.3f} kJ/mol"
+            )
+
+    plt.tight_layout()
+    buffer = io.BytesIO()
+    plt.savefig(buffer, format="png")
+    buffer.seek(0)
+    img_as_bytes = buffer.read()
+
+    return all_dGs, all_errs, img_as_bytes
+
+
+def estimate_relative_free_energy(mol_a, mol_b, core, ff, host_config, n_frames=1000, prefix=""):
+    """
+    Estimate relative free energy between mol_a and mol_b. Molecules should be aligned to each
+    other and within the host environment.
+
+    Parameters
+    ----------
+    mol_a: Chem.Mol
+        initial molecule
+
+    mol_b: Chem.Mol
+        target molecule
+
+    core: list of 2-tuples
+        atom_mapping of atoms in mol_a into atoms in mol_b
+
+    ff: ff.Forcefield
+        Forcefield to be used for the system
+
+    host_config: HostConfig
+        Configuration for the host system.
+
+    n_frames: int
+        number of samples to generate for each lambda windows
```

Comment:
"generate for endstate lambda windows"

Since you filter them below

Reply:
... where each sample is after 750 steps of MD.

Reply:
willfix

Reply:
to clarify, this is saying that each window runs for n_frames*1000 steps of MD. It's independent of how many frames we store etc. I'll add a note in the return type that only first and last windows are currently returned.